azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import abc
+import six
+from collections import deque
+from . import pipeline_ops
+from . import pipeline_events
+from . import mqtt_topic_helper
+from azure.iot.device.iothub.models import Message
+from azure.iot.device.common.transport.mqtt.mqtt_provider import MQTTProvider
+from azure.iot.device.common.transport import constant
+
+logger = logging.getLogger(__name__)
+
+# TODO: standardize on {} instead of %s
+# TODO: standardize logs with {}({}) prefix
```

Comment:
Potentially also consider to perhaps use {keywords} instead of just {}. It's not necessary, but it could be more readable. I don't do this too often, but it is nice when there's no surrounding context to draw from.

e.g.

```
"{name}({operation}) handling".format(name=self.name, operation=op.name)
```